contracts/deploy/utils/index.ts

```diff
@@ -0,0 +1,100 @@
+import { HardhatRuntimeEnvironment } from "hardhat/types";
+import inquirer from "inquirer";
+
+import ERC1967Proxy from "@openzeppelin/contracts/build/contracts/ERC1967Proxy.json";
+import UUPSUpgradeable from "@openzeppelin/contracts-upgradeable/build/contracts/UUPSUpgradeable.json";
+
+interface DeployOpts {
+  initializer?: string;
+}
+
+export function getProxyName(name: string): string {
+  return `Proxy__${name}`;
+}
+
+export async function deployUUPSProxiedContract(
+  hre: HardhatRuntimeEnvironment,
+  deployer: string,
+  name: string,
+  args: any[],
+  opts?: DeployOpts
+) {
+  const { ethers, upgrades, deployments, network } = hre;
+  const { deploy, getOrNull } = deployments;
+  const initializer = opts?.initializer ?? "initialize";
+  const deployerSigner = await ethers.getSigner(deployer);
+
+  const Factory = await ethers.getContractFactory(name);
+  // Validate if implementation is upgradeable
+  await upgrades.validateImplementation(Factory, { kind: "uups" });
+
+  const existingDeployment = await getOrNull(getProxyName(name));
+  if (existingDeployment) {
+    console.log(
+      `Trying to upgrade proxy deployed at ${existingDeployment.address}`
+    );
+    const { doUpgrade } = await inquirer.prompt([
+      {
+        type: "confirm",
+        name: "doUpgrade",
+        message: "Do you want to upgrade the proxy?",
+        default: false,
+      },
+    ]);
+    if (!doUpgrade) {
```

Comment:
nit `Aborting upgradation` looks like a better message.

Reply:
the entire deployment process is aborted